generic: mtk_eth_soc: add support for flow-control settings

dangowrt opened the pull request:
Add patch implementing operations to get and set flow-control link parameters of mtk_eth_soc to Linux 6.1 and Linux 6.6 generic patchset. This allows setting and retrieving current flow-control settings via `ethtool` or via `ubus` and UCI network config.

Example:
```
root@bpi-r4:/# ethtool -a eth2
Pause parameters for eth2:
Autonegotiate:  on
RX:             off
TX:             off
RX negotiated: off
TX negotiated: off
```

```
root@OpenWrt:/# ubus call network.device status '{"name":"eth2"}'
{
        "external": false,
        "present": true,
        "type": "Network device",
        "up": true,
        "carrier": true,
        "auth_status": false,
        "link-advertising": [
                "10baseT-H",
                "10baseT-F",
                "100baseT-H",
                "100baseT-F",
                "1000baseT-F",
                "1000baseX-F"
        ],
        "link-partner-advertising": [
                "10baseT-H",
                "10baseT-F",
                "100baseT-H",
                "100baseT-F",
                "1000baseT-F"
        ],
        "link-supported": [
                "10baseT-H",
                "10baseT-F",
                "100baseT-H",
                "100baseT-F",
                "1000baseT-F",
                "1000baseX-F"
        ],
        "speed": "1000F",
        "autoneg": true,
        "flow-control": {
                "autoneg": true,
                "supported": [
                        "pause",
                        "asym_pause"
                ],
                "link-advertising": [

                ],
                "link-partner-advertising": [
                        "pause"
                ],
                "negotiated": [

                ]
        },
        "hw-tc-offload": true,
        "devtype": "ethernet",
        "mtu": 1500,
        "mtu6": 1500,
        "macaddr": "36:26:4f:c2:de:13",
        "txqueuelen": 1000,
        "ipv6": false,
        "ip6segmentrouting": false,
        "promisc": false,
        "rpfilter": 0,
        "acceptlocal": false,
        "igmpversion": 0,
        "mldversion": 0,
        "neigh4reachabletime": 30000,
        "neigh6reachabletime": 30000,
        "neigh4gcstaletime": 60,
        "neigh6gcstaletime": 60,
        "neigh4locktime": 100,
        "dadtransmits": 1,
        "multicast": true,
        "sendredirects": true,
        "drop_v4_unicast_in_l2_multicast": false,
        "drop_v6_unicast_in_l2_multicast": false,
        "drop_gratuitous_arp": false,
        "drop_unsolicited_na": false,
        "arp_accept": false,
        "gro": true,
        "statistics": {
                "collisions": 0,
                "rx_frame_errors": 0,
                "tx_compressed": 0,
                "multicast": 0,
                "rx_length_errors": 0,
                "tx_dropped": 0,
                "rx_bytes": 54843533,
                "rx_missed_errors": 0,
                "tx_errors": 0,
                "rx_compressed": 0,
                "rx_over_errors": 0,
                "tx_fifo_errors": 0,
                "rx_crc_errors": 0,
                "rx_packets": 93039,
                "tx_heartbeat_errors": 0,
                "rx_dropped": 0,
                "tx_aborted_errors": 0,
                "tx_packets": 41172,
                "rx_errors": 0,
                "tx_bytes": 13169372,
                "tx_window_errors": 0,
                "rx_fifo_errors": 0,
                "tx_carrier_errors": 0
        }
}
```

dangowrt commented:
@the2masters Yes, AQR112C is kinda special as it performs rate-adaptation using pause frames (ie. the link between MAC and PHY is always using 2500Base-X, no matter the external speed). Also the driver for AQR112C is kinda minimalistic and may not handle flow-control settings correctly at this point. I also got this device here and will check later.

Spudz76 commented:
Reads status okay on kernel 6.6 with Linksys EA7300-V1 (mt7621)
```
root@wonky2:~# ethtool -a eth0
Pause parameters for eth0:
Autonegotiate:  off
RX:             on
TX:             on
```

Won't allow any changes though, could be normal for this board.
```
root@wonky2:~# ethtool -A eth0 autoneg on
netlink error: Not supported
root@wonky2:~# ethtool -A eth0 tx off
netlink error: Not supported
root@wonky2:~# ethtool -A eth0 rx off
netlink error: Not supported
```

Spudz76 commented:
I figured out that eth0 is the core device but the actual ports are named `lanX`, and all commands work properly on those.

Headcrabed commented:
Sorry to bother, but what happened to this patch? Seems not only aqr phys but also 88x3310 phy need this patch to work correctly.

dangowrt committed:
generic: 6.6: mtk_eth_soc: add support for flow-control settings

Add patch implementing operations to get and set flow-control link
parameters of mtk_eth_soc via ethtool.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Spudz76 reviewed (approved):
Works on Linksys EA7300v1 (ramips/mt7621)

rsalvaterra commented:
Working perfectly on my Redmi AC2100, thanks, @dangowrt!

`[   27.716082] mtk_soc_eth 1e100000.ethernet wan: Link is Up - 1Gbps/Full - flow control off`

glassd00r commented:
Tested working on bpi-r4 (mt7988a)
